### PR TITLE
Adds flaming shots and Flaming Moes

### PR DESCRIPTION
**Pull request by Citinited**
(sorry Homer ;_;)

This PR implements / tweaks the following:
- You can now set shots alight using something hot! Flaming shots give off light and will probably burn you if you are clumsy and try to drink them. 
- Adds the Flaming Moe, a somewhat alcoholic beverage that requires fire to prepare.

**Obligatory pictures!**

Notes:

- This only works with shot glasses. Drinks that are in normal glasses will be unable to be set alight, as I cannot be arsed to add flaming sprites for every drink we have.
- This also only works with shotglasses filled with ethanol. Suggestions for how to implement it for stuff like napalm and welding fuel without adding a var on every reagent would be appreciated.

:cl:
add: Adds the Flaming Moe cocktail.
add: Shot glasses can now be set alight using something hot.
/:cl:

### CODE REVIEW

**Comment by DesolateG** [comment]
It's a Flaming Homer!

**Comment by scrubmcnoob** [comment]
*let's get waassstedddddd*

also that cigarette thing is really nice

File: code/game/objects/items/weapons/cigs.dm [hunk 1]
```diff
@@ -103,6 +90,8 @@ LIGHTERS ARE IN LIGHTERS.DM
 				explosion(user.loc, -1, 0, 2, 3, 0, flame_range = 2)
 			F.charges--
 
+	else if(is_hot(W) > 600)
```
Comment (Fox-McCloud) [inline code/game/objects/items/weapons/cigs.dm]: Save this for a different PR. `is_hot` is a terrible proc that violates OOP; likewise, the cig lighting needs to be made more OOP, as well.

File: code/__HELPERS/unsorted.dm [hunk 2]
```diff
@@ -1196,6 +1196,12 @@ var/global/list/common_tools = list(
 			return 0
 	if(istype(W, /obj/item/device/assembly/igniter))
 		return 1000
+	if(istype(W, /obj/item/weapon/reagent_containers/food/drinks/drinkingglass/shotglass))
```
Comment (Fox-McCloud) [inline code/__HELPERS/unsorted.dm]: This is probably going to create problems. I don't recommend doing this.

This proc is primarily used in `attackby` and some other things; overriding reagent container default behavior is generrrrallyy not a good thing.

As a prime example, trying to dump one reagent container into another, with this, will probably heat up the reagent container you attempt to dump it into.

**Comment by Fox-McCloud** [comment]
Goo idea, but I think this could be implemented a bit differently.

It may be better to just have it so that shotglasses override their parent in terms of how they function with the fire system.

Just have it so that instead of it catching fire and clearing reagents.....it checks for if there's ethanol then sets it alight; the fire system will handle it becoming extinguished and everything else about it.

It's far less snowflakey and still has most of the intended effect.

**Comment by Citinited** [comment]
Hmm, I didn't consider doing it that way. It's not fantastic code, so I'll see if I can rewrite what I have to incorporate that.

I'll also need to play around with it so that most functionality I want to add is there.

File: code/game/objects/items/weapons/cigs.dm (hunk 1 above)
Comment (Citinited) [inline code/game/objects/items/weapons/cigs.dm]: Noted. RIP cigarettes lighting on shots :P

File: code/__HELPERS/unsorted.dm (hunk 2 above)
Comment (Citinited) [inline code/__HELPERS/unsorted.dm]: It's on fire though, that's to be expected ;)

File: code/modules/reagents/chemistry/reagents/alcohol.dm [hunk 3]
```diff
@@ -424,6 +424,18 @@
 	drink_desc = "Does... does this mean that Arthur and Ford are on the station? Oh joy."
 	taste_message = "the number fourty two"
 
+/datum/reagent/consumable/ethanol/flaming_homer
+	name = "Flaming Moe"
+	id = "flamingmoe"
+	description = "This seems like a bunch of various alcohols thrown together, along with what looks like prescription medicine..."
```
Comment (SamHPurp) [inline code/modules/reagents/chemistry/reagents/alcohol.dm]: various is superfluous with `bunch of`, I'd drop the `various`, grammatically.

Comment (SamHPurp) [inline code/modules/reagents/chemistry/reagents/alcohol.dm]: Similarly `seems like` and `looks like`, grammatically I'd change the `what looks like` to `appears to be`.

File: code/game/objects/items/weapons/cigs.dm (hunk 1 above)
Comment (Citinited) [inline code/game/objects/items/weapons/cigs.dm]: I think I'll refactor `is_hot` after this PR to just be object flags; that way it's nice and simple

File: code/modules/reagents/chemistry/reagents/alcohol.dm (hunk 3 above)
Comment (Citinited) [inline code/modules/reagents/chemistry/reagents/alcohol.dm]: Yeah, I'll change this around

**Comment by shazbot194** [comment]
One though on how to add flames to every normal glass, have a flame overlay and something like 2-3 flame sprites, and have it so each drink picks a flame then offsets it based on where the opening for the drink is, plus this system could have any sized fire you want even with tall glasses that normally would only have a couple pixels to work with.

File: code/modules/reagents/chemistry/reagents/alcohol.dm (hunk 3 above)
Comment (taukausanake) [inline code/modules/reagents/chemistry/reagents/alcohol.dm]: Wait, you can toast drinks? That's awesome, now we just need to learn how to set water on fire

Comment (Citinited) [inline code/modules/reagents/chemistry/reagents/alcohol.dm]: It's only the flaming moe that has toasting. It's basically only flavour text.

*mmm... flavour text...*

**Comment by Citinited** [comment]
@shazbot194 It's more difficult than just "add flame overlay, reposition". Many of our drinks have completely different rim sizes, and some have animations above the rim that the fire overlay would muck around with.

File: code/modules/food_and_drinks/drinks/drinks/shotglass.dm [hunk 4]
```diff
@@ -20,9 +28,78 @@
...
+
+/obj/item/weapon/reagent_containers/food/drinks/drinkingglass/shotglass/proc/isShotFlammable()
+	var/datum/reagent/R = reagents.get_master_reagent()
+	if(istype(R, /datum/reagent/consumable/ethanol))
+		var/datum/reagent/consumable/ethanol/A = R
+		if(A.volume >= 5 && A.alcohol_perc >= 0.35) //Only an approximation to if something's flammable but it will do
+			return TRUE
+
+/obj/item/weapon/reagent_containers/food/drinks/drinkingglass/shotglass/fire_act(global_overlay = FALSE, var/silent = FALSE)
+	if(!isShotFlammable() || burn_state == ON_FIRE) //You can't light a shot that's not flammable!
```
Comment (Citinited) [inline code/modules/food_and_drinks/drinks/drinks/shotglass.dm]: Not sure why `fire_act` still runs through the whole proc proc even when the object is on fire. Overriden here

**Comment by Citinited** [comment]
PR updated to use existing procs for stuff. The cigarette thing has been axed and will probably be put into another PR; flaming shots heating up the shot glass has also been removed.

File: code/game/objects/items.dm [hunk 5]
```diff
@@ -202,7 +202,7 @@ var/global/image/fire_overlay = image("icon" = 'icons/goonstation/effects/fire.d
 		to_chat(user, msg)
 
 
-/obj/item/attack_hand(mob/user as mob)
+/obj/item/attack_hand(mob/user as mob, var/pickupfireoverride = FALSE)
```
Comment (Crazylemon64) [inline code/game/objects/items.dm]: Please do not include the `var/` prefix when defining proc arguments!

File: code/modules/food_and_drinks/drinks/drinks/shotglass.dm [hunk 6]
```diff
@@ -20,9 +28,78 @@
...
+	extinguish(silent = TRUE)
+
+/obj/item/weapon/reagent_containers/food/drinks/drinkingglass/shotglass/proc/isShotFlammable()
+	var/datum/reagent/R = reagents.get_master_reagent()
+	if(istype(R, /datum/reagent/consumable/ethanol))
+		var/datum/reagent/consumable/ethanol/A = R
+		if(A.volume >= 5 && A.alcohol_perc >= 0.35) //Only an approximation to if something's flammable but it will do
+			return TRUE
+
+/obj/item/weapon/reagent_containers/food/drinks/drinkingglass/shotglass/fire_act(global_overlay = FALSE, var/silent = FALSE)
```
Comment (Crazylemon64) [inline code/modules/food_and_drinks/drinks/drinks/shotglass.dm]: Please do not include the var/ prefix when defining proc arguments!

File: code/modules/food_and_drinks/drinks/drinks/shotglass.dm [hunk 7]
```diff
@@ -20,9 +28,78 @@
...
+	if(!isShotFlammable() || burn_state == ON_FIRE) //You can't light a shot that's not flammable!
+		return
+	..()
+	set_light(light_intensity, null, light_color)
+	if(!silent)
+		visible_message("<span class = 'notice'>[src] begins to burn with a blue hue!</span>")
+	update_icon()
+
+/obj/item/weapon/reagent_containers/food/drinks/drinkingglass/shotglass/extinguish(var/silent = FALSE)
```
Comment (Crazylemon64) [inline code/modules/food_and_drinks/drinks/drinks/shotglass.dm]: Please do not include the var/ prefix when defining proc arguments!

**Comment by Citinited** [comment]
@Crazylemon64 Done

**Pull request merged**